submissions/views/xqueue.py

```diff
@@ -0,0 +1,296 @@
+            log.error("Invalid reply from pull-grader: request.data: %s",
+                      request.data)
+            return Response(
+                self.compose_reply(False, 'Incorrect reply format'),
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            external_grader = ExternalGraderDetail.objects.select_for_update(
+                                                                        nowait=True).get(submission__id=submission_id)
+        except ExternalGraderDetail.DoesNotExist:
+            log.error(
+                "Grader submission_id refers to nonexistent entry in Submission DB: "
+                "grader: %s, submission_key: %s, score_msg: %s",
+                submission_id,
+                submission_key,
+                score_msg
+            )
+            return Response(
+                self.compose_reply(False, 'Submission does not exist'),
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+        if not external_grader.pullkey or submission_key != external_grader.pullkey:
+            log.error(f"Invalid pullkey: submission key from xwatcher {submission_key} "
+                      f"and submission key stored {external_grader.pullkey} are different")
+            return Response(
+                self.compose_reply(False, 'Incorrect key for submission'),
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        try:
+            log.info("Attempting to set_score...")
+            set_score(str(external_grader.submission.uuid),
+                      points_earned,
+                      external_grader.points_possible
+                      )
+            external_grader.update_status('retired', score_msg)
+            log.info("Successfully updated submission score for submission %s", submission_id)
+
+        except (SubmissionNotFoundError, DatabaseError, SubmissionInternalError) as e:
+            log.exception("Error when executing set_score: %s", type(e).__name__)
+            external_grader.update_status("failed", score_msg)
+
+        return Response(self.compose_reply(success=True, content=''))
+
+    def validate_grader_reply(self, external_reply):
+        """
+        Validate the format of external grader reply.
+
+        Returns:
+            tuple: (is_valid, submission_id, submission_key, score_msg)
+        """
+        fail = (False, -1, '', '', '')
+
+        if not isinstance(external_reply, dict):
+            return fail
+
+        try:
+            header = external_reply['xqueue_header']
+            score_msg = external_reply['xqueue_body']
+        except KeyError:
+            return fail
+
+        try:
+            header_dict = json.loads(header)
+        except (TypeError, ValueError):
+            return fail
+
+        try:
+            score = json.loads(score_msg)
+            points_earned = score.get("score")
+        except (TypeError, ValueError):
+            return fail
```

Comment:
Adding error logging to this validation would be nice, since the only thing people running this in production will currently see is that the replies are bad, but not why they're bad.